### DISCUSSION TITLE
Reading list

### DISCUSSION THREAD

**Discussion by gvanrossum** [faster-cpython/ideas#123] [General]
Please add your favorite references!

Two things on HHVM (2018, so fairly recent):
- video: https://www.youtube.com/watch?v=P58g--fPNDM&ab_channel=PLDI2018
- paper: https://research.fb.com/wp-content/uploads/2018/04/hhvm-jit-a-profile-guided-region-based-compiler-for-php-and-hack.pdf

Mark Shannon writings:
- HotPy (2): https://sites.google.com/site/makingcpythonfast/
- A faster CPython: https://github.com/markshannon/faster-cpython
- Ph.D thesis: http://theses.gla.ac.uk/2975/

Victor Stinner writings:
- Faster CPython: https://faster-cpython.readthedocs.io/index.html
- FAT Python: https://faster-cpython.readthedocs.io/fat_python.html

**Comment by lewurm** [comment]
Inline Caching:
* Inline Caching meets Quickening [2009]: http://www.complang.tuwien.ac.at/kps09/pdfs/brunthaler.pdf
* BPO issue for that: https://bugs.python.org/issue14757
* Explaining JavaScript VMs in JavaScript - Inline Caches [2012]: https://mrale.ph/blog/2012/06/03/explaining-js-vms-in-js-inline-caches.html
* Max Bernstein on inline caching [2021]: https://bernsteinbear.com/blog/inline-caching/

Hidden classes:
* Fast properties in v8 [2017]: https://v8.dev/blog/fast-properties
* `super` property access improvement in v8 [2021]: https://v8.dev/blog/fast-super

**Comment by markshannon** [comment]
Analyses of Python runtime overhead:

* Python Interpreter Performance Deconstructed [2014]: https://www.complang.tuwien.ac.at/gergo/papers/dyla14.pdf
* Quantitative Overhead Analysis for Python [unpublished?]: https://cpb-us-w2.wpmucdn.com/sites.coecis.cornell.edu/dist/7/89/files/2016/08/iiswc18-revised-1fjavbb.pdf
* Profiling CPython at Instagram: https://instagram-engineering.com/profiling-cpython-at-instagram-89d4cbeeb898

Interpreter dispatch:

* Branch Prediction and the Performance of Interpreters -Don’t Trust Folklore [2015]: https://hal.inria.fr/hal-01100647/document

**Comment by markshannon** [comment]
[Producing Wrong Data Without Doing Anything Obviously Wrong!](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.163.8395)

**Comment by ericsnowcurrently** [comment]
"USING PYTHON FOR MODEL INFERENCE IN DEEP LEARNING", DeVito et al (Facebook AI Research)

https://arxiv.org/abs/2104.00254

The most relevant aspects, for our project, are their coverage of a popular class of Python workloads (PyTorch) and their approach to CPU parallelism using CPython.

Overall, they cover a lot of interesting details that extend to many parts of Python core development and to the Python packaging ecosystem.  Most notably:

* a popular class of Python workloads (PyTorch)
* embedding CPython
* running multiple interpreters in the same process with CPU parallelism
* interpreter isolation
* solutions for package dependency resolution
* packaging PyTorch models for efficiency

**Comment by gvanrossum** [comment]
A very detailed post on perf improvements for .NET 6:

https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/

**Comment by gvanrossum** [comment]
Copy-and-Patch Compilation:

https://arxiv.org/pdf/2011.13127.pdf

**Comment by ericsnowcurrently** [comment]
From #213: https://www.sciencedirect.com/science/article/abs/pii/S0167642321001520 (paywalled).

**Comment by gvanrossum** [comment]
Simple and Effective Type Check Removal through Lazy Basic Block Versioning

https://arxiv.org/pdf/1411.0352.pdf

**Comment by itamaro** [comment]
Writeup about the function inliner in Cinder JIT

https://engineering.fb.com/2022/05/02/open-source/cinder-jits-instagram/

**Comment by lpereira** [comment]
Threaded Code Generation with a Meta-Tracing JIT Compiler

https://arxiv.org/pdf/2106.12496.pdf
